crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,236 @@
+                                        err
+                                    );
+                                    tokio::time::sleep(Duration::from_secs(1)).await;
+                                    metrics.checkpoint_exec_errors.inc();
+                                }
+
+                                (seq_num, next_checkpoint_clone)
+                            }));
+
+                            // Last checkpoint of epoch
+                            if let Some(next_committee) = next_checkpoint.next_epoch_committee() {
+                                // Unwrap ok because reconfig failure is unrecoverable
+                                self.reconfig(next_committee).await.unwrap();
+                            }
+                        }
+                    }
+                    // In this case, messages in the mailbox have been overwritten
+                    // as a result of lagging too far behind. We can simply continue the loop,
+                    // as the next call to recv() should return the overwritten checkpoint and we
+                    // will attempt to execute all checkpoints from (highest_executed_checkpoint + 1)
+                    // up to the result of the next recv() call.
+                    Err(RecvError::Lagged(num_skipped)) => {
+                        warn!(
+                            "Checkpoint Execution Recv channel overflowed {:?} messages",
+                            num_skipped,
+                        );
+                        self.metrics
+                            .checkpoint_exec_recv_channel_overflow
+                            .inc_by(num_skipped);
+                    }
+                    Err(RecvError::Closed) => {
+                        info!("Checkpoint Execution Sender (StateSync) closed channel");
+                        break;
+                    }
+                },
+            }
+        }
+    }
+
+    async fn reconfig(&self, _next_epoch_committee: &[(AuthorityName, StakeUnit)]) -> SuiResult {
+        // TODO(william) call into Reconfig
+        Ok(())
+    }
+
+    fn update_lag_metric(&self, highest_executed: CheckpointSequenceNumber) -> SuiResult {
+        let highest_synced = self
+            .checkpoint_store
+            .get_highest_synced_checkpoint_seq_number()?
+            .unwrap_or(0);
+        // TODO may want to consider asserting >= 0
+        let diff = highest_synced - highest_executed;
+        self.metrics.checkpoint_exec_lag.set(diff as i64);
+        Ok(())
+    }
+}
+
+pub async fn execute_checkpoint(
+    checkpoint: VerifiedCheckpoint,
+    authority_state: Arc<AuthorityState>,
+    checkpoint_store: Arc<CheckpointStore>,
+) -> SuiResult<Vec<TransactionEffects>> {
+    let effects = if let Some(checkpoint_contents) = checkpoint_store
+        .get_checkpoint_contents(&checkpoint.content_digest())
+        .map_err(SuiError::from)?
```

Comment:
nit: you don't need `map_err` here

Reply:
Why not? This method returns `TypedStoreError` on error, which is not a `SuiError` type... would it be coerced?

Reply:
the `?` sigil will implicitly try to do a conversion as it desugars into effectively this:
 ```
let foo = match failable_op() {
    Ok(foo) => foo,
    Err(e) => return Err(e.into()),
};
```

Reply:
nit: this could still be fixed.